ambry-clustermap/src/main/java/com/github/ambry/clustermap/InstanceConfigToDataNodeConfigAdapter.java

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright 2020 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *
+ */
+
+package com.github.ambry.clustermap;
+
+import com.github.ambry.config.ClusterMapConfig;
+import java.util.Objects;
+import org.apache.helix.HelixManager;
+import org.apache.helix.api.listeners.InstanceConfigChangeListener;
+import org.apache.helix.model.InstanceConfig;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static com.github.ambry.clustermap.ClusterMapUtils.*;
+
+
+/**
+ * An implementation of {@link DataNodeConfigSource} that converts between {@link InstanceConfig}s received by an
+ * {@link InstanceConfigChangeListener} and {@link DataNodeConfig}s.
+ */
+public class InstanceConfigToDataNodeConfigAdapter implements DataNodeConfigSource {
+  private static final Logger LOGGER = LoggerFactory.getLogger(InstanceConfigToDataNodeConfigAdapter.class);
+  private final HelixManager helixManager;
+  private final ClusterMapConfig clusterMapConfig;
+
+  /**
+   * @param helixManager the {@link HelixManager} to use as the source of truth for {@link InstanceConfig}s.
+   * @param clusterMapConfig the {@link ClusterMapConfig} to use.
+   */
+  public InstanceConfigToDataNodeConfigAdapter(HelixManager helixManager, ClusterMapConfig clusterMapConfig) {
+    this.helixManager = helixManager;
+    this.clusterMapConfig = clusterMapConfig;
+  }
+
+  @Override
+  public void addServerConfigChangeListener(DataNodeConfigChangeListener listener) throws Exception {
+    helixManager.addInstanceConfigChangeListener((InstanceConfigChangeListener) (instanceConfigs, context) -> {
+      Iterable<DataNodeConfig> dataNodeConfigs =
+          () -> instanceConfigs.stream().map(this::convert).filter(Objects::nonNull).iterator();
+      listener.onDataNodeConfigChange(dataNodeConfigs);
```

Comment:
Nice written to use observe pattern so that when DataNodeConfig changes state, the clusterChangeHandler are notified and updated through onDataNodeConfigChange!